globus_action_provider_tools/flask/middleware/cloudwatch_metrics.py

```diff
@@ -0,0 +1,239 @@
+
+import json
+import logging
+import typing as t
+from datetime import datetime, timedelta
+
+from flask import Response, g
+
+log = logging.getLogger("cloudwatch-emf-middleware")
+
+
+class CloudWatchMetricEMFLogger:
+    """
+    Middleware to emit CloudWatch Metrics detailing action provider usage via
+      the CloudWatch EMF Format.
+    https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html
+
+    Metric Structure
+    ================
+
+    Aggregate
+    ---------
+      Namespace: {supplied_namespace}
+      Dimensions:
+        ActionProvider: {supplied_action_provider_name}
+
+    Route-Specific
+    --------------
+      Namespace: {supplied_namespace}
+      Dimensions:
+        ActionProvider: {supplied_action_provider_name}
+        Route: "run" | "resume" | "status" | ...
+
+    Included Metrics:
+     * Count - The total number API requests in a given period.
+     * 2XXs - The number of successful responses returned in a given period.
+     * 4XXs - The number of client-side errors captured in a given period.
+     * 5XXs - The number of server-side errors captured in a given period.
+     * RequestLatency - The number of milliseconds between the request being received
+        and the response being sent.
+    """
+
+    def __init__(
+        self, namespace: str, action_provider_name: str, log_level: int | None = None
+    ):
+        """
+        :param namespace: Custom CloudWatch Namespace target
+        :param action_provider_name: Action Provider Name to be used in metric dimension
+           sets
+        :param log_level: Optional log level to use when emitting metrics. If None,
+           metrics will be printed to stdout instead of logged.
+        """
+        self._namespace = namespace
+        self._action_provider_name = action_provider_name
+        self._log_level = log_level
+
+    def before_request(self):
+        g.request_start_time = datetime.now()
+
+    def after_request(self, response: Response):
+        if hasattr(g, "route_type") and hasattr(g, "request_start_time"):
+            self.emit_route_metrics(
+                route_name=g.route_type,
+                request_latency=datetime.now() - g.request_start_time,
+                response_status=response.status_code,
+            )
+        return response
+
+    def teardown_request(self, error: BaseException | None):
+        # If a request errors mid-handling, after_request handlers will not be called,
+        #   so we need to emit metrics for errors separately here
+        if error:
+            if hasattr(g, "route_type") and hasattr(g, "request_start_time"):
+                status_code = 500
+                if hasattr(error, "code"):
+                    status_code = error.code
+                self.emit_route_metrics(
+                    route_name=g.route_type,
+                    request_latency=datetime.now() - g.request_start_time,
```

Comment:
This may not matter terribly much in practice, but would it be better to use something like `time.perf_counter` or `time.monotonic` that is designed for this type of usage?

Reply:
Sure I can switch; I'm not super familiar with the distinction between various ways of figuring out the current time in python.

Reply:
Seems like `perf_counter` is the most designed for this use case, so switching to that.

Reply:
This seems to have solved the problem.

Probably was running fast enough that something about integer conversion in datetime.now() & timedelta was truncating the sub-ms runtime in certain circumstances to 0.

Now it's emitted as a floating point value instead & CI run verifies it's > 0. Thanks for the recommendation.